File: frontend/wiretap-ui/src/api/captureEvents.ts
// src/api/captureEvents.ts
//
// Events — a moment (or a span) and a note, owned by the stored capture.
// A local capture keeps its own; a WireTAP Backend profile's database keeps
// them on the gateway. Both answer through the same Rust commands.

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export interface CaptureEvent {
  id: string;
  timestampUs: number;
  /** 0 marks an instant; anything more is a span from `timestampUs`. */
  durationUs: number;
  note: string;
  createdAtUs: number;
  updatedAtUs: number;
}

export type EventOwner =
  | { kind: "capture"; capture_id: string }
  | { kind: "backend"; profile_id: string };

/** The editable part of an event. */
export type EventDraft = Pick<CaptureEvent, "timestampUs" | "durationUs" | "note">;

export function listCaptureEvents(owner: EventOwner): Promise<CaptureEvent[]> {
  return invoke<CaptureEvent[]>("capture_events_list", { owner });
}

export function addCaptureEvent(owner: EventOwner, draft: EventDraft): Promise<CaptureEvent> {
  return invoke<CaptureEvent>("capture_events_add", { owner, ...draft });
}

export function updateCaptureEvent(owner: EventOwner, id: string, draft: EventDraft): Promise<CaptureEvent> {
  return invoke<CaptureEvent>("capture_events_update", { owner, id, ...draft });
}

export function deleteCaptureEvent(owner: EventOwner, id: string): Promise<void> {
  return invoke("capture_events_delete", { owner, id });
}

/** Fires in every window after any mutation, with the owner that changed. */
export function onCaptureEventsChanged(callback: (owner: EventOwner) => void): Promise<UnlistenFn> {
  return listen<EventOwner>("capture-events-changed", (event) => callback(event.payload));
}

File: crates/wiretap-app/src/capture_events.rs
// crates/wiretap-app/src/capture_events.rs
//
// Events — a moment (or a span) and a note, owned by the stored capture: a
// local SQLite capture keeps its own rows, a WireTAP Backend profile's
// database keeps them on the gateway. Every mutation broadcasts
// `capture-events-changed` with the owner so each window refreshes.

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter};

use crate::{apiclient, capture_db};

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureEvent {
    pub id: String,
    pub timestamp_us: i64,
    pub duration_us: i64,
    pub note: String,
    pub created_at_us: i64,
    pub updated_at_us: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum EventOwner {
    Capture { capture_id: String },
    Backend { profile_id: String },
}

const CHANGED_EVENT: &str = "capture-events-changed";

#[tauri::command]
pub async fn capture_events_list(app: AppHandle, owner: EventOwner) -> Result<Vec<CaptureEvent>, String> {
    match &owner {
        EventOwner::Capture { capture_id } => capture_db::list_capture_events(capture_id),
        EventOwner::Backend { profile_id } => apiclient::events_list(&app, profile_id).await,
    }
}

#[tauri::command]
pub async fn capture_events_add(
    app: AppHandle,
    owner: EventOwner,
    timestamp_us: i64,
    duration_us: i64,
    note: String,
) -> Result<CaptureEvent, String> {
    let event = match &owner {
        EventOwner::Capture { capture_id } => {
            capture_db::insert_capture_event(capture_id, timestamp_us, duration_us, &note)
        }
        EventOwner::Backend { profile_id } => {
            apiclient::events_add(&app, profile_id, timestamp_us, duration_us, &note).await
        }
    }?;
    let _ = app.emit(CHANGED_EVENT, &owner);
    Ok(event)
}

#[tauri::command]
pub async fn capture_events_update(
    app: AppHandle,
    owner: EventOwner,
    id: String,
    timestamp_us: i64,
    duration_us: i64,
    note: String,
) -> Result<CaptureEvent, String> {
    let event = match &owner {
        EventOwner::Capture { capture_id } => {
            capture_db::update_capture_event(capture_id, &id, timestamp_us, duration_us, &note)
        }
        EventOwner::Backend { profile_id } => {
            apiclient::events_update(&app, profile_id, &id, timestamp_us, duration_us, &note).await
        }
    }?;
    let _ = app.emit(CHANGED_EVENT, &owner);
    Ok(event)
}

#[tauri::command]
pub async fn capture_events_delete(app: AppHandle, owner: EventOwner, id: String) -> Result<(), String> {
    match &owner {
        EventOwner::Capture { capture_id } => capture_db::delete_capture_event(capture_id, &id),
        EventOwner::Backend { profile_id } => apiclient::events_delete(&app, profile_id, &id).await,
    }?;
    let _ = app.emit(CHANGED_EVENT, &owner);
    Ok(())
}
